File: MedicalAppointments_UPDATED/project/ProjectFinal/backend/MedicalAppointments.API/Services/SlotService.cs
using MedicalAppointments.API.DTOs.Slot;
using MedicalAppointments.API.Hubs;
using MedicalAppointments.API.Infrastructure;
using MedicalAppointments.API.Models;
using Microsoft.AspNetCore.SignalR;
using Microsoft.EntityFrameworkCore;

namespace MedicalAppointments.API.Services;

public interface ISlotService
{
    Task<TimeSlotResponseDto> CreateAsync(CreateSlotDto dto, int doctorId);
    Task<IEnumerable<TimeSlotResponseDto>> CreateBulkAsync(CreateBulkSlotsDto dto, int doctorId);
    Task BlockAsync(int slotId, int doctorId);
    Task UnblockAsync(int slotId, int doctorId);
    Task<IEnumerable<TimeSlotResponseDto>> GetWeekSlotsAsync(int doctorId, DateOnly weekStart);
    Task<bool> IsWeekFullAsync(int doctorId, DateOnly weekStart);
}

public class SlotService : ISlotService
{
    private readonly AppDbContext          _db;
    private readonly IHubContext<SlotHub>  _hub;
    private readonly ILogger<SlotService>  _logger;

    public SlotService(
        AppDbContext db,
        IHubContext<SlotHub> hub,
        ILogger<SlotService> logger)
    {
        _db     = db;
        _hub    = hub;
        _logger = logger;
    }

    public async Task<TimeSlotResponseDto> CreateAsync(CreateSlotDto dto, int doctorId)
    {
        if (dto.EndTime <= dto.StartTime)
            throw new ArgumentException("L'heure de fin doit être après l'heure de début.");

        if (dto.StartTime < DateTime.UtcNow)
            throw new ArgumentException("Impossible de créer un créneau dans le passé.");

        await CheckOverlapAsync(doctorId, dto.StartTime, dto.EndTime);

        var slot = new TimeSlot
        {
            DoctorId  = doctorId,
            StartTime = dto.StartTime,
            EndTime   = dto.EndTime,
            Status    = SlotStatus.Available
        };

        _db.TimeSlots.Add(slot);

        _db.AuditLogs.Add(new AuditLog
        {
            UserId     = doctorId,
            Action     = "CREATE_SLOT",
            EntityType = "TimeSlot",
            EntityId   = 0,
            Detail     = $"{dto.StartTime:dd/MM/yyyy HH:mm} – {dto.EndTime:HH:mm}"
        });

        await _db.SaveChangesAsync();

        await _hub.Clients.All.SendAsync("SlotAdded", MapToDto(slot));

        return MapToDto(slot);
    }

    public async Task<IEnumerable<TimeSlotResponseDto>> CreateBulkAsync(
        CreateBulkSlotsDto dto, int doctorId)
    {
        var created = new List<TimeSlot>();
        var current = dto.WeekStartDate.ToDateTime(TimeOnly.MinValue);
        var weekEnd = current.AddDays(7);

        while (current < weekEnd)
        {
            if (dto.WorkDays.Contains(current.DayOfWeek))
            {
                var start = current.Date
                    .AddHours(dto.SlotStartTime.Hour)
                    .AddMinutes(dto.SlotStartTime.Minute);
                var dayEnd = current.Date
                    .AddHours(dto.SlotEndTime.Hour)
                    .AddMinutes(dto.SlotEndTime.Minute);

                while (start.AddMinutes(dto.SlotDurationMinutes) <= dayEnd)
                {
                    var end = start.AddMinutes(dto.SlotDurationMinutes);

                    bool overlap = await _db.TimeSlots.AnyAsync(t =>
                        t.DoctorId == doctorId &&
                        t.Status != SlotStatus.Cancelled &&
                        t.StartTime < end &&
                        t.EndTime   > start);

                    if (!overlap)
                    {
                        var slot = new TimeSlot
                        {
                            DoctorId  = doctorId,
                            StartTime = start,
                            EndTime   = end,
                            Status    = SlotStatus.Available
                        };
                        _db.TimeSlots.Add(slot);
                        created.Add(slot);
                    }

                    start = end;
                }
            }
            current = current.AddDays(1);
        }

        await _db.SaveChangesAsync();

        // Notifier en temps réel
        foreach (var slot in created)
            await _hub.Clients.All.SendAsync("SlotAdded", MapToDto(slot));

        return created.Select(MapToDto);
    }

    public async Task BlockAsync(int slotId, int doctorId)
    {
        var slot = await _db.TimeSlots
            .FirstOrDefaultAsync(t => t.Id == slotId && t.DoctorId == doctorId)
            ?? throw new KeyNotFoundException("Créneau introuvable.");

        if (slot.Status == SlotStatus.Reserved)
            throw new InvalidOperationException(
                "Impossible de bloquer un créneau déjà réservé.");

        slot.Status    = SlotStatus.Blocked;
        slot.UpdatedAt = DateTime.UtcNow;

        _db.AuditLogs.Add(new AuditLog
        {
            UserId     = doctorId,
            Action     = "BLOCK_SLOT",
            EntityType = "TimeSlot",
            EntityId   = slotId
        });

        await _db.SaveChangesAsync();

        await _hub.Clients.All.SendAsync("SlotStatusChanged", new
        {
            slotId      = slot.Id,
            status      = "Blocked",
            isClickable = false
        });
    }

    public async Task UnblockAsync(int slotId, int doctorId)
    {
        var slot = await _db.TimeSlots
            .FirstOrDefaultAsync(t => t.Id == slotId && t.DoctorId == doctorId)
            ?? throw new KeyNotFoundException("Créneau introuvable.");

        if (slot.Status != SlotStatus.Blocked)
            throw new InvalidOperationException("Ce créneau n'est pas bloqué.");

        slot.Status    = SlotStatus.Available;
        slot.UpdatedAt = DateTime.UtcNow;

        await _db.SaveChangesAsync();

        await _hub.Clients.All.SendAsync("SlotStatusChanged", new
        {
            slotId      = slot.Id,
            status      = "Available",
            isClickable = true
        });
    }

    public async Task<IEnumerable<TimeSlotResponseDto>> GetWeekSlotsAsync(
        int doctorId, DateOnly weekStart)
    {
        var start = weekStart.ToDateTime(TimeOnly.MinValue);
        var end   = start.AddDays(7);

        var slots = await _db.TimeSlots
            .Where(t => t.DoctorId == doctorId
                     && t.StartTime >= start
                     && t.StartTime <  end)
            .OrderBy(t => t.StartTime)
            .ToListAsync();

        return slots.Select(MapToDto);
    }

    public async Task<bool> IsWeekFullAsync(int doctorId, DateOnly weekStart)
    {
        var start = weekStart.ToDateTime(TimeOnly.MinValue);
        var end   = start.AddDays(7);

        return !await _db.TimeSlots.AnyAsync(t =>
            t.DoctorId == doctorId &&
            t.StartTime >= start &&
            t.StartTime <  end &&
            t.Status == SlotStatus.Available);
    }

    // ─── Helpers privés ────────────────────────────────────────────────────
    private async Task CheckOverlapAsync(int doctorId, DateTime start, DateTime end)
    {
        bool overlap = await _db.TimeSlots.AnyAsync(t =>
            t.DoctorId == doctorId &&
            t.Status != SlotStatus.Cancelled &&
            t.StartTime < end &&
            t.EndTime   > start);

        if (overlap)
            throw new InvalidOperationException("Ce créneau chevauche un créneau existant.");
    }

    private static TimeSlotResponseDto MapToDto(TimeSlot t) =>
        new(t.Id, t.StartTime, t.EndTime,
            t.Status.ToString(),
            t.Status == SlotStatus.Available);
}

File: appointment-service/src/main/java/com/hospital/appointment/service/SlotService.java
package com.hospital.appointment.service;

import com.hospital.appointment.dto.slot.CreateBulkSlotsDto;
import com.hospital.appointment.dto.slot.CreateSlotDto;
import com.hospital.appointment.dto.slot.HospitalSlotResponseDto;
import com.hospital.appointment.dto.slot.TimeSlotResponseDto;
import com.hospital.appointment.entity.AuditLog;
import com.hospital.appointment.entity.SlotStatus;
import com.hospital.appointment.entity.TimeSlot;
import com.hospital.appointment.entity.Appointment;
import com.hospital.appointment.exception.BusinessException;
import com.hospital.appointment.exception.NotFoundException;
import com.hospital.appointment.repository.AuditLogRepository;
import com.hospital.appointment.repository.TimeSlotRepository;
import com.hospital.appointment.websocket.SlotEventPublisher;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

/**
 * Service de gestion des créneaux (TimeSlot).
 * Migré depuis Services/SlotService.cs (.NET).
 */
@Service
@RequiredArgsConstructor
@Slf4j
public class SlotService {

    private final TimeSlotRepository timeSlotRepository;
    private final AuditLogRepository auditLogRepository;
    private final SlotEventPublisher events;

    private static final DateTimeFormatter SLOT_FMT =
            DateTimeFormatter.ofPattern("dd/MM/yyyy HH:mm", Locale.FRENCH);
    private static final DateTimeFormatter TIME_FMT =
            DateTimeFormatter.ofPattern("HH:mm");

    // ─────────────────────────────────────────────────────────────────
    //  CRÉATION D'UN CRÉNEAU UNIQUE
    // ─────────────────────────────────────────────────────────────────
    @Transactional
    public TimeSlotResponseDto create(CreateSlotDto dto, Integer doctorId) {
        if (!dto.endTime().isAfter(dto.startTime())) {
            throw new IllegalArgumentException("L'heure de fin doit être après l'heure de début.");
        }
        if (dto.startTime().isBefore(LocalDateTime.now())) {
            throw new IllegalArgumentException("Impossible de créer un créneau dans le passé.");
        }

        checkOverlap(doctorId, dto.startTime(), dto.endTime());

        TimeSlot slot = TimeSlot.builder()
                .doctorId(doctorId)
                .startTime(dto.startTime())
                .endTime(dto.endTime())
                .status(SlotStatus.Available)
                .build();
        slot = timeSlotRepository.save(slot);

        auditLogRepository.save(AuditLog.builder()
                .userId(doctorId)
                .action("CREATE_SLOT")
                .entityType("TimeSlot")
                .entityId(slot.getId())
                .detail("%s – %s".formatted(
                        dto.startTime().format(SLOT_FMT),
                        dto.endTime().format(TIME_FMT)))
                .build());

        TimeSlotResponseDto result = mapToDto(slot, null);
        events.publishSlotAdded(result);
        return result;
    }

    // ─────────────────────────────────────────────────────────────────
    //  CRÉATION EN MASSE (semaine entière)
    // ─────────────────────────────────────────────────────────────────
    @Transactional
    public List<TimeSlotResponseDto> createBulk(CreateBulkSlotsDto dto, Integer doctorId) {
        List<TimeSlot> created = new ArrayList<>();
        LocalDate currentDate = dto.weekStartDate();
        LocalDate weekEnd = currentDate.plusDays(7);

        while (currentDate.isBefore(weekEnd)) {
            if (dto.workDays().contains(currentDate.getDayOfWeek())) {
                LocalDateTime dayStart = currentDate.atTime(dto.slotStartTime());
                LocalDateTime dayEnd = currentDate.atTime(dto.slotEndTime());

                LocalDateTime cursor = dayStart;
                while (cursor.plusMinutes(dto.slotDurationMinutes()).isBefore(dayEnd.plusSeconds(1))) {
                    LocalDateTime slotEnd = cursor.plusMinutes(dto.slotDurationMinutes());

                    boolean overlap = timeSlotRepository.existsOverlap(doctorId, cursor, slotEnd);
                    if (!overlap) {
                        TimeSlot slot = TimeSlot.builder()
                                .doctorId(doctorId)
                                .startTime(cursor)
                                .endTime(slotEnd)
                                .status(SlotStatus.Available)
                                .build();
                        created.add(timeSlotRepository.save(slot));
                    }
                    cursor = slotEnd;
                }
            }
            currentDate = currentDate.plusDays(1);
        }

        List<TimeSlotResponseDto> dtos = new ArrayList<>();
        for (TimeSlot slot : created) {
            TimeSlotResponseDto d = mapToDto(slot, null);
            dtos.add(d);
            events.publishSlotAdded(d);
        }
        return dtos;
    }

    // ─────────────────────────────────────────────────────────────────
    //  BLOQUER / DÉBLOQUER
    // ─────────────────────────────────────────────────────────────────
    @Transactional
    public void block(Integer slotId, Integer doctorId) {
        TimeSlot slot = timeSlotRepository.findByIdAndDoctorId(slotId, doctorId)
                .orElseThrow(() -> new NotFoundException("Créneau introuvable."));

        if (slot.getStatus() == SlotStatus.Reserved) {
            throw new BusinessException("Impossible de bloquer un créneau réservé.");
        }

        slot.setStatus(SlotStatus.Blocked);
        slot.setUpdatedAt(LocalDateTime.now());
        timeSlotRepository.save(slot);

        events.publishSlotStatusChanged(slot.getId(), "Blocked", false);
    }

    @Transactional
    public void unblock(Integer slotId, Integer doctorId) {
        TimeSlot slot = timeSlotRepository.findByIdAndDoctorId(slotId, doctorId)
                .orElseThrow(() -> new NotFoundException("Créneau introuvable."));

        if (slot.getStatus() != SlotStatus.Blocked) {
            throw new BusinessException("Ce créneau n'est pas bloqué.");
        }

        slot.setStatus(SlotStatus.Available);
        slot.setUpdatedAt(LocalDateTime.now());
        timeSlotRepository.save(slot);

        events.publishSlotStatusChanged(slot.getId(), "Available", true);
    }

    // ─────────────────────────────────────────────────────────────────
    //  CALENDRIER MÉDECIN
    // ─────────────────────────────────────────────────────────────────
    @Transactional(readOnly = true)
    public List<TimeSlotResponseDto> getWeekSlots(Integer doctorId, LocalDate weekStart) {
        LocalDateTime start = weekStart.atStartOfDay();
        LocalDateTime end = start.plusDays(7);

        return timeSlotRepository.findWeekSlotsByDoctor(doctorId, start, end)
                .stream()
                .map(t -> mapToDto(t, t.getAppointment()))
                .toList();
    }

    @Transactional(readOnly = true)
    public boolean isWeekFull(Integer doctorId, LocalDate weekStart) {
        LocalDateTime start = weekStart.atStartOfDay();
        LocalDateTime end = start.plusDays(7);
        return !timeSlotRepository.hasAvailableSlot(doctorId, start, end, SlotStatus.Available);
    }

    // ─────────────────────────────────────────────────────────────────
    //  CRÉNEAUX HOSPITALIERS (DoctorId NULL)
    // ─────────────────────────────────────────────────────────────────
    @Transactional(readOnly = true)
    public List<HospitalSlotResponseDto> getHospitalSlots(Integer serviceId, LocalDate weekStart) {
        LocalDateTime start = weekStart.atStartOfDay();
        LocalDateTime end = start.plusDays(7);

        return timeSlotRepository.findHospitalWeekSlots(serviceId, start, end)
                .stream()
                .map(t -> new HospitalSlotResponseDto(
                        t.getId(),
                        t.getStartTime(),
                        t.getEndTime(),
                        t.getStatus().name(),
                        t.getStatus() == SlotStatus.Available,
                        t.getServiceId() != null ? t.getServiceId() : 0,
                        t.getService() != null ? t.getService().getName() : ""))
                .toList();
    }

    @Transactional(readOnly = true)
    public boolean isHospitalWeekFull(Integer serviceId, LocalDate weekStart) {
        LocalDateTime start = weekStart.atStartOfDay();
        LocalDateTime end = start.plusDays(7);
        return !timeSlotRepository.hasAvailableHospitalSlot(serviceId, start, end, SlotStatus.Available);
    }

    // ─── Helpers privés ──────────────────────────────────────────────
    private void checkOverlap(Integer doctorId, LocalDateTime start, LocalDateTime end) {
        if (timeSlotRepository.existsOverlap(doctorId, start, end)) {
            throw new BusinessException("Ce créneau chevauche un créneau existant.");
        }
    }

    /**
     * Mapping vers le DTO de réponse, avec infos patient si réservé (FIX Bug 3).
     */
    static TimeSlotResponseDto mapToDto(TimeSlot t, Appointment appt) {
        String patientName = null;
        String patientEmail = null;
        Integer apptId = null;
        if (appt != null && appt.getPatient() != null) {
            patientName = appt.getPatient().getFullName();
            patientEmail = appt.getPatient().getEmail();
            apptId = appt.getId();
        }
        return new TimeSlotResponseDto(
                t.getId(),
                t.getStartTime(),
                t.getEndTime(),
                t.getStatus().name(),
                t.getStatus() == SlotStatus.Available,
                patientName,
                patientEmail,
                apptId);
    }
}
